apps/scan/src/main/scala/org/lfdecentralizedtrust/splice/scan/store/bulk/ZstdGroupedWeight.scala

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) 2024 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package org.lfdecentralizedtrust.splice.scan.store.bulk
+
+import com.github.luben.zstd.ZstdDirectBufferCompressingStreamNoFinalizer
+import org.apache.pekko.stream.{Attributes, FlowShape, Inlet, Outlet}
+import org.apache.pekko.stream.stage.{GraphStage, GraphStageLogic, InHandler, OutHandler}
+import org.apache.pekko.util.ByteString
+
+import java.nio.ByteBuffer
+import java.util.concurrent.atomic.AtomicReference
+
+/** A Pekko GraphStage that zstd-compresses a stream of bytestrings, and splits the output into zstd objects of size (minWeight + delta).
+  * Somewhat similar to Pekko's built-in GroupedWeight, but outputs valid zstd compressed objects.
+  */
+case class ZstdGroupedWeight(minSize: Long) extends GraphStage[FlowShape[ByteString, ByteString]] {
```

Comment:
preStart and postStop can probably be used to init and cleanup the compressingStream (in GraphStageLogic) (I should have made this comment on the GraphStageLogic further down).
Especially postStop important to always close the compressingStream.